master/internal/hpimportance/manager.go

```diff
@@ -0,0 +1,284 @@
+	case ExperimentCreated:
+		m.state[msg.ID] = stateRecord{
+			lastResult: time.Now(),
+		}
+	case ExperimentPaused:
+		m.triggerDefaultWork(ctx, msg.ID)
+	case ExperimentProgress:
+		var state stateRecord
+		var ok bool
+		if state, ok = m.state[msg.ID]; !ok {
+			state = stateRecord{}
+		}
+		if msg.Progress-state.lastProgress > minPercent &&
+			time.Since(state.lastResult) > minPause {
+			m.triggerDefaultWork(ctx, msg.ID)
+		}
+	case workStarted:
+		hpi, err := m.db.GetHPImportance(msg.experimentID)
+		if err != nil {
+			ctx.Log().Errorf("error retrieving hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+		metricData := getMetricHPImportance(hpi, msg.metricName, msg.metricType)
+		metricData.Status = InProgress
+		setMetricHPImportance(&hpi, metricData, msg.metricName, msg.metricType)
+		err = m.db.SetHPImportance(msg.experimentID, hpi)
+		if err != nil {
+			ctx.Log().Errorf("error writing hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+	case workFailed:
+		hpi, err := m.db.GetHPImportance(msg.experimentID)
+		if err != nil {
+			ctx.Log().Errorf("error retrieving hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+		metricData := getMetricHPImportance(hpi, msg.metricName, msg.metricType)
+		metricData.Status = Failed
+		metricData.Error = msg.err
+		setMetricHPImportance(&hpi, metricData, msg.metricName, msg.metricType)
+		err = m.db.SetHPImportance(msg.experimentID, hpi)
+		if err != nil {
+			ctx.Log().Errorf("error writing hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+	case workCompleted:
+		m.state[msg.experimentID] = stateRecord{
+			lastResult:   time.Now(),
+			lastProgress: m.state[msg.experimentID].lastProgress,
+		}
+		hpi, err := m.db.GetHPImportance(msg.experimentID)
+		if err != nil {
+			ctx.Log().Errorf("error retrieving hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+		metricData := getMetricHPImportance(hpi, msg.metricName, msg.metricType)
+		metricData.ExperimentProgress = msg.progress
+		metricData.HpImportance = msg.results
+		switch metricData.Status {
+		case Pending:
+			// Do nothing - this means another startWork message was already sent
+		case InProgress:
+			metricData.Status = Complete
+		default:
+			ctx.Log().Warnf("work was completed for a metric with an unexpected state")
+		}
+		setMetricHPImportance(&hpi, metricData, msg.metricName, msg.metricType)
+		err = m.db.SetHPImportance(msg.experimentID, hpi)
+		if err != nil {
+			ctx.Log().Errorf("error writing hyperparameter importance state: %s", err.Error())
+			return nil
+		}
+	default:
+		ctx.Log().Errorf("unknown message received by hyperparameter importance manager: %v!",
+			ctx.Message())
+	}
+	return nil
+}
+
+func (m *manager) getChild(ctx *actor.Context, experimentID int) *actor.Ref {
+	var result *actor.Ref
+	/*
+		Currently each experiment gets their own actor, that will persist even after the work
```

Comment:
Had a chat with Ryan about this a while back because I was concerned about it and didn't see a super clean & easy solution. Apparently this is already the situation with experiment actors (for reasons) and the actors are intended to be super-light-weight and numerous, so not a huge concern. But still I'd like to clean it up. We may want to change from 1-actor-per-experiment to a sort of thread-pool model, but there may be other implications in that decision coming soon once we pick a random forest implementation, and understand the resource constraints better (i.e. is it better for experiments to be isolated from each other, or to have a hard cap on total resource usage at any one time), etc. So if it's cool with you, I'd like to leave that for now.